pkg/controller/servicemesh/controlplane/reconciler.go

```diff
@@ -341,6 +352,25 @@ func (r *controlPlaneInstanceReconciler) Reconcile(ctx context.Context) (result
 	return
 }
 
+// finds the first SMCP in the given namespace by creationTimestamp
+func getEarliestSMCPInNamespace(ctx context.Context, cl client.Client, ns string) (*v2.ServiceMeshControlPlane, error) {
+	smcpList := &v2.ServiceMeshControlPlaneList{}
+	if err := cl.List(ctx, smcpList, client.InNamespace(ns)); err != nil {
+		return nil, err
+	}
+
+	var earliest *v2.ServiceMeshControlPlane
+	for _, smcp := range smcpList.Items {
+		if earliest == nil ||
+			smcp.CreationTimestamp.Before(&earliest.CreationTimestamp) ||
+			(smcp.CreationTimestamp == earliest.CreationTimestamp && smcp.Name < earliest.Name) {
+			goPitfallWorkaround := smcp
+			earliest = &goPitfallWorkaround
```

Comment:
Since `smcp` is not a pointer, but a value, why do you need this temporary variable?

Reply:
Because the smcp variable gets reused in each step of the for loop (it's like being declared outside of the for loop). So, storing the address of this variable in `earliest` will result in `earliest` always pointing to the last smcp in the loop.

For example, let's say you have 2 smcps that you're iterating over. In the first step, you set `earliest` to the address of the smcp variable (which at that point holds the 1st smcp). At the beginning of the second step, the value in the smcp variable is overridden with the 2nd smcp. Even though you don't update `earliest` in the second step, it will point to the a variable that holds the value of the 2nd smcp.

So, to have `earliest` point to the correct value, you need to create a new variable (`goPitfallWorkaround` in this case) and get its address. 

The go authors plan to change this in a future release.

Reply:
See discussion here: https://github.com/golang/go/discussions/56010

Reply:
Yes, but I think you could avoid it by defining `earliest` as `v2.ServiceMeshControlPlane` instead of `*v2.ServiceMeshControlPlane`.

Reply:
True, but `getEarliestSMCPInNamespace()` should return `nil`, when no SMCPs are found. I mean, I could modify the calling code to correctly handle an empty SMCP, but somehow I find it semantically incorrect to return an empty SMCP when no SMCPs are found. But maybe I'm wrong in thinking this way.

Reply:
Ok, I forgot about returning nil... In such a case, it's absolutely reasonable to define this variable as a pointer.
But to be honest, I don't like the name `goPitfallWorkaround`. For me it's an expected behavior and I wouldn't call it Go's pitfall, so I would prefer the following alternatives:
1. Initialize the pointer and set its value:
```suggestion
			earliest = &v2.ServiceMeshControlPlane{}
			*earliest = smcp
```
2. Rename `goPitfallWorkaround` to `earliestTmp`:
```suggestion
			earliestTmp := smcp
			earliest = &earliestTmp
```

Reply:
The reason to explicitly call this out as a pitfall is because it is a pitfall. It's probably the #1 pitfall in go. If a person is not aware of this, they might accidentally change the code, thinking they are simplifying it, when they are actually breaking it. 

In both of your examples, it's not clear why the code isn't simply `earliest = &smcp`. 

By explicitly calling out the pitfall, no-one will dare to change the code without fully understanding why it's written like it is. The variable name serves as a reminder.

Reply:
Ok, fine. You don't have to change it, but you won't convince me that this is a pitfall. This is just the nature of pointers. I think that the approach with a "tmp" variable is more clear, because it's a typical notation in algorithmic handbooks in C, so most people should quickly see what's going on.

> If a person is not aware of this, they might accidentally change the code

I would say that if a person is not aware of pointers, they should not touch Go. Actually changing the code without understanding how pointers work is safe, because you wrote tests.

Reply:
But this isn't about knowing how pointers work. It's about knowing how loops are implemented in Go. And as you can see in the discussion, virtually everyone runs into this problem (which is why they're going to change how loops work):

> ### A decade of experience shows the cost of the current semantics
> I [talked at Gophercon once](https://go.dev/blog/toward-go2#explaining-problems) about how we need agreement about the existence of a problem before we move on to solutions. When we examined this issue in the run up to Go 1, it did not seem like enough of a problem. The general consensus was that it was annoying but not worth changing.
>
> Since then, **I suspect every Go programmer in the world has made this mistake in one program or another. I certainly have done it repeatedly over the past decade, despite being the one who argued for the current semantics and then implemented them. (Sorry!)**

I agree that tests would catch any breaking changes, but when a person runs into a failing test, they still need to figure out what's causing it to fail. This unnecessarily wastes time and money. It's always better to explicitly point out pitfalls like this.

Reply:
Leaving aside naming, which is always a never-ending debate, what I think could be good to leave in this piece of code is an explanation **why** we do what we do. A short comment with reference to this interesting thread about for-loop semantics will not only serve us as backup memory of why this decision was taken, but help any newcomer who stumbles upon this code base to spend some extra effort understanding the intricacies and give some extra care when touching it. Just my two cents.